lineapy/graph_reader/artifact_collection.py

```diff
@@ -110,92 +110,186 @@ def __init__(
+
+        return inter_session_taskgraph
+
+    def create_inter_artifact_taskgraph(
+        self, dependencies: TaskGraphEdge
+    ) -> TaskGraph:
+        """
+        Returns TaskGraph for dependencies between "Artifacts".
+
+        Nodes correspond to nodecollections.
+
+        Edges between artifacts in different sessions are replaced
+        so that all the dependencies pointing to the target artifact instead point to
+        its "root" tasks in the session which are source tasks that can be used to generate the artifact.
+
+        Session 1: A -> B
+        Session 2: C -> D
+
+        If a user specifies a dependency from B -> D, the inter artifact taskgraph will
+        instead draw a dependency from B -> C to ensure B runs before all the intra session
+        dependencies of D.
+
+        Note: These additional edges prevent bugs as the user maybe be unable to specify dependencies
+        to automatically generated components.
+        """
+        inter_artifact_taskgraph = TaskGraph(nodes=[], edges={})
+
+        # add subgraph for each session_artifact
+        for sa in self.session_artifacts.values():
+            inter_artifact_taskgraph.graph = nx.compose(
+                inter_artifact_taskgraph.graph,
+                sa.nodecollection_dependencies.graph,
+            )
+
+        art_name_to_session_id: Dict[str, str] = {}
+        for artifact in self.target_artifacts:
+            art_name_to_session_id[slugify(artifact.name)] = str(
+                artifact._session_id
+            )
+
+        self._validate_dependencies_used(
+            inter_artifact_taskgraph, dependencies
+        )
+
+        for to_artname, from_artname_set in dependencies.items():
+            for from_artname in from_artname_set:
+                from_session_id, to_session_id = (
+                    art_name_to_session_id[from_artname],
+                    art_name_to_session_id[to_artname],
+                )
+
+                if from_session_id != to_session_id:
+                    # this edge needs to be replaced, all the dependencies of target artifact in the same session
+                    # must be run after the source artifact is created
+
+                    to_graph = self.session_artifacts[
+                        LineaID(to_session_id)
+                    ].nodecollection_dependencies
+
+                    to_art_ancestors = nx.ancestors(
+                        to_graph.graph, source=to_artname
+                    )
+
+                    for to_source_node in to_graph.source_nodes:
+                        if to_source_node in to_art_ancestors:
+                            inter_artifact_taskgraph.graph.add_edge(
+                                from_artname, to_source_node
+                            )
+                else:
+                    inter_artifact_taskgraph.graph.add_edge(
+                        from_artname, to_artname
+                    )
```

Comment:
This is adding intra-session edges based on user-provided dependency info, which may not be correct. We may skip it and rely on `SessionArtifacts.nodecollection_dependencies` for intra-session edges.

Reply:
Ok I can remove this else block.